source/redx/os/win_utils.cpp

```diff
@@ -17,49 +17,80 @@
-      ValueName[0] = '\0';
-
-      if (RegEnumValueW(
-        hKey, i++, ValueName, &cchValueName, NULL, NULL, NULL, NULL)
-        != ERROR_SUCCESS)
-      {
-        break;
-      }
-
-      std::filesystem::path p(ValueName);
-
-      if (p.filename() == exename && std::filesystem::exists(p))
-      {
-        ret = p;
-        break;
-      }
+    std::optional<std::filesystem::path> find_exe_path(std::string_view exename) {
+        // Computer\HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Compatibility Assistant\Store
+        // Computer\HKEY_CURRENT_USER\SOFTWARE\Classes\Local Settings\Software\Microsoft\Windows\Shell\MuiCache
+
+        std::optional<std::filesystem::path> ret;
+
+        HKEY hKey;
+
+        if (RegOpenKeyExA(
+            HKEY_CURRENT_USER,
+            "SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\AppCompatFlags\\Compatibility Assistant\\Store",
+            0, KEY_READ, &hKey)
+            == ERROR_SUCCESS) 
+        {
+            WCHAR ValueName[MAX_VALUE_NAME];
+            DWORD cchValueName = MAX_VALUE_NAME;
+
+            DWORD i = 0;
+            while (1) 
+            {
+                cchValueName = MAX_VALUE_NAME;
+                ValueName[0] = '\0';
+
+                if (RegEnumValueW(
+                    hKey, i++, ValueName, &cchValueName, NULL, NULL, NULL, NULL)
+                    != ERROR_SUCCESS) 
+                {
+                    RegCloseKey(hKey);
+                    break;
+                }
+
+                std::filesystem::path p(ValueName);
+
+                if (p.filename() == exename && std::filesystem::exists(p)) 
+                {
+                    RegCloseKey(hKey);
+                    return p;
+                }
+            }
+        }
+
+        if (RegOpenKeyExA(
+            HKEY_CURRENT_USER,
+            "SOFTWARE\\Classes\\Local Settings\\Software\\Microsoft\\Windows\\Shell\\MuiCache",
+            0, KEY_READ, &hKey)
+            == ERROR_SUCCESS) 
+        {
+            WCHAR ValueName[MAX_VALUE_NAME];
+            DWORD cchValueName = MAX_VALUE_NAME;
+
+            DWORD i = 0;
+            while (1) 
+            {
+                cchValueName = MAX_VALUE_NAME;
+                ValueName[0] = '\0';
+
+                if (RegEnumValueW(hKey, i++, ValueName, &cchValueName, NULL, NULL, NULL, NULL) != ERROR_SUCCESS) 
+                {
+                    RegCloseKey(hKey);
+                    break;
+                }
+                std::filesystem::path p(ValueName);
+
+                // Keys here are in format: Cyberpunk2077.exe.XXX
+                // -Cyberpunk2077.exe.ApplicationCompany
+                // -Cyberpunk2077.exe.FriendlyAppName
+                if (p.filename().string().find(exename) != std::string::npos && std::filesystem::exists(p.remove_filename().append(exename))) 
+                {
+                    RegCloseKey(hKey);
+                    return p.string();

```

Comment:
returning p.string() may not be what you intended.
```cpp
if (p.filename().string().find(exename) != std::string::npos) {
  p = p.remove_filename().append(exename);
  if (std::filesystem::exists(p) {
    return p;
  }
}
```

Reply:
Oh, right, it should be `return p`. Also, I don't think it's necessary to do: `p = p.remove_filename().append(exename);`

How about the following? `p.remove_filename().append(exename);`

Since `remove_filename()` and `append()` both return a reference:

![image](https://user-images.githubusercontent.com/63656305/195245663-754560c8-cee7-4b82-9f32-cc03d638496d.png)
[Link to Doc](https://en.cppreference.com/w/cpp/filesystem/path/remove_filename)

So…

```
if (p.filename().string().find(exename) != std::string::npos)
{
  p.remove_filename().append(exename);
  if (std::filesystem::exists(p)) 
  {
    RegCloseKey(hKey);
    return p;
  }
}

...

return std::nullopt; // unrelated, but saw a bug that would cause it to throw errors when using has_value()
```

P.S.: I'll also update the code to conform to this project's indentation standards.